sdk/python/kfp/client/auth.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2022 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import os
+from webbrowser import open_new_tab
+
+import google.auth
+import google.auth.app_engine
+import google.auth.compute_engine.credentials
+import google.auth.iam
+import google.oauth2.credentials
+import google.oauth2.service_account
+import requests
+import requests_toolbelt.adapters.appengine
+from google.auth.transport.requests import Request
+
+IAM_SCOPE = 'https://www.googleapis.com/auth/iam'
+OAUTH_TOKEN_URI = 'https://www.googleapis.com/oauth2/v4/token'
+LOCAL_KFP_CREDENTIAL = os.path.expanduser('~/.config/kfp/credentials.json')
+
+
+def get_gcp_access_token():
+    """Gets GCP access token for the current Application Default Credentials.
+
+    If not set, returns None. For more information, see
+    https://cloud.google.com/sdk/gcloud/reference/auth/application-default/print-access-token
+    """
+    token = None
+    try:
+        creds, project = google.auth.default(
+            scopes=["https://www.googleapis.com/auth/cloud-platform"])
+        if not creds.valid:
+            auth_req = Request()
+            creds.refresh(auth_req)
+        if creds.valid:
+            token = creds.token
+    except Exception as e:
+        logging.warning('Failed to get GCP access token: %s', e)
+    return token
+
+
+def get_auth_token(client_id, other_client_id, other_client_secret):
+    """Gets auth token from default service account or user account."""
+    if os.path.exists(LOCAL_KFP_CREDENTIAL):
+        # fetch IAP auth token using the locally stored credentials.
+        with open(LOCAL_KFP_CREDENTIAL, 'r') as f:
+            credentials = json.load(f)
+        if client_id in credentials:
+            return id_token_from_refresh_token(
+                credentials[client_id]['other_client_id'],
+                credentials[client_id]['other_client_secret'],
+                credentials[client_id]['refresh_token'], client_id)
+    if other_client_id is None or other_client_secret is None:
```

Comment:
nit / not your code:

This could be refactored to make it a little bit clearer what is returned if the predicate `if other_client_id is None or other_client_secret is None` is met:

```python
    if other_client_id is None or other_client_secret is None:
        # fetch IAP auth token: service accounts
        return get_auth_token_from_sa(client_id)
    # fetch IAP auth token: user account
    # Obtain the ID token for provided Client ID with user accounts.
    # Flow: get authorization code -> exchange for refresh token -> obtain
    # and return ID token
    refresh_token = get_refresh_token_from_client_id(
        other_client_id, other_client_secret)
    credentials = {}
    if os.path.exists(LOCAL_KFP_CREDENTIAL):
        with open(LOCAL_KFP_CREDENTIAL, 'r') as f:
            credentials = json.load(f)
    credentials[client_id] = {}
    credentials[client_id]['other_client_id'] = other_client_id
    credentials[client_id]['other_client_secret'] = other_client_secret
    credentials[client_id]['refresh_token'] = refresh_token
    # TODO: handle the case when the refresh_token expires, which only
    # happens if the refresh_token is not used once for six months.
    if not os.path.exists(os.path.dirname(LOCAL_KFP_CREDENTIAL)):
        os.makedirs(os.path.dirname(LOCAL_KFP_CREDENTIAL))
    with open(LOCAL_KFP_CREDENTIAL, 'w') as f:
        json.dump(credentials, f)
    return id_token_from_refresh_token(other_client_id,
                                            other_client_secret, refresh_token,
                                            client_id)
```

Reply:
This file and client.py as well may have many refactoring opportunities, let alone the pylint errors. 
Will leave those for later PRs-- feel empowered to propose refactoring PRs as well :)